File: EticaretJava/app/src/main/java/com/example/eticaretjava/adapter/CategoryAdapter.java
package com.example.eticaretjava.adapter;


import android.view.LayoutInflater;
import android.view.ViewGroup;


import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;


import com.example.eticaretjava.databinding.ItemCategoryBinding;
import com.example.eticaretjava.model.Category;


public class CategoryAdapter
        extends ListAdapter<Category.CategoryDto, CategoryAdapter.VH> {


    public interface OnSelected {
        void onSelected(Category.CategoryDto item);
    }


    private final OnSelected onSelected;
    private Integer selectedId = null;


    public CategoryAdapter(OnSelected onSelected) {
        super(DIFF);
        this.onSelected = onSelected;
    }


    public void setSelected(Integer id) {
        selectedId = id;
        notifyDataSetChanged();
    }


    static class VH extends RecyclerView.ViewHolder {
        ItemCategoryBinding b;


        VH(ItemCategoryBinding b) {
            super(b.getRoot());
            this.b = b;
        }
    }


    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemCategoryBinding b = ItemCategoryBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);
        return new VH(b);
    }


    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        Category.CategoryDto item = getItem(position);


        holder.b.chip.setText(item.name);
        holder.b.chip.setChecked(item.id == (selectedId != null ? selectedId : -1));


        holder.b.chip.setOnClickListener(v -> {
            selectedId = (selectedId != null && selectedId == item.id)
                    ? null : item.id;


            onSelected.onSelected(
                    selectedId == null ? null : item
            );


            notifyDataSetChanged();
        });
    }


    private static final DiffUtil.ItemCallback<Category.CategoryDto> DIFF =
            new DiffUtil.ItemCallback<Category.CategoryDto>() {
                @Override
                public boolean areItemsTheSame(@NonNull Category.CategoryDto o,
                                               @NonNull Category.CategoryDto n) {
                    return o.id == n.id;
                }


                @Override
                public boolean areContentsTheSame(@NonNull Category.CategoryDto o,
                                                  @NonNull Category.CategoryDto n) {
                    return false;
                }
            };
}

File: EticaretKotlin/app/src/main/java/com/example/eticaretkotlin/adapters/CategoryAdapter.kt
package com.example.eticaretkotlin.adapters


import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.eticaretkotlin.databinding.ItemCategoryBinding
import com.example.eticaretkotlin.model.CategoryDto

class CategoryAdapter(
    private val onSelected: (CategoryDto?) -> Unit
) : ListAdapter<CategoryDto, CategoryAdapter.VH>(DIFF) {

    private var selectedId: Int? = null

    fun setSelected(id: Int?) {
        selectedId = id
        notifyDataSetChanged()
    }

    inner class VH(val b: ItemCategoryBinding) : RecyclerView.ViewHolder(b.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        val b = ItemCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return VH(b)
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        val item = getItem(position)
        holder.b.chip.text = item.name
        holder.b.chip.isChecked = (item.id == selectedId)
        holder.b.chip.setOnClickListener {
            selectedId = if (selectedId == item.id) null else item.id
            onSelected(if (selectedId == null) null else item)
            notifyDataSetChanged()
        }
    }

    companion object {
        private val DIFF = object : DiffUtil.ItemCallback<CategoryDto>() {
            override fun areItemsTheSame(o: CategoryDto, n: CategoryDto) = o.id == n.id
            override fun areContentsTheSame(o: CategoryDto, n: CategoryDto) = o == n
        }
    }
}
